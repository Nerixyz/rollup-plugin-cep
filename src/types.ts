import { XMLWriterOptions } from 'xmlbuilder2/lib/interfaces';

export interface Options {
  configPath: string;
  outputPath: string;
  debugOutputPath: string;
  emitDebug: boolean;
  writerOptions?: XMLWriterOptions;
}

export interface Config {
  manifestVersion?: Version;
  bundleId: string;
  bundleName?: string;
  bundleVersion?: Version;
  extensions: Extension[];
  executionEnvironment: ExecutionEnvironment;
  author?: string;
  contact?: { mailto?: string; value?: string };
  legal?: string | { href?: string; value: string };
  abstract?: string | { href?: string; value: string };
}

export interface ExecutionEnvironment {
  hostList: Record<string, Version>;
  localeList?: string[];
  requiredRuntime?: Version;
}

export type Vec2 = [number | string | undefined, number | string | undefined];
export type Version = string | string[] | number | number[];

export interface Extension {
  id: string;
  version?: Version;
  dependencies?: Record<string, string>;
  hostList?: string[];
  dispatchInfos: ExtensionDispatchInfo[];
  debug?: Record<string, number | string>;
}

export interface ExtensionDispatchInfo {
  mainPath?: string;
  scriptPath?: string | { value: string; engine?: string };
  commandLineParameters?: string[];
  autoVisible?: boolean;
  startOn?: string[];
  type?: 'Panel' | 'ModalDialog' | 'Modeless' | 'Custom' | 'Embedded' | 'Dashboard';
  menu?: string | { value: string; placement?: string };
  screenPercentage?: Vec2;
  size?: Vec2;
  maxSize?: Vec2;
  minSize?: Vec2;
  icons?: Record<'Normal' | 'Disabled' | 'RollOver' | 'DarkNormal' | 'DarkRollOver', string>;
  host?: string;
}
