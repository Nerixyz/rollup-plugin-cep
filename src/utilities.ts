import { Config, Vec2, Version } from './types';
import {resolve} from 'path';

export function makeVersion(version: Version): string {
  if (!Array.isArray(version)) return version.toString();
  return version.join('.');
}

export function importConfig(path: string): Config {
  let mod = require(resolve(path));
  if (mod.default) mod = mod.default;
  if (typeof mod === 'function') return mod();
  return mod;
}

export function optAttrToXml<T extends string>(
  obj:
    | undefined
    | string
    | ({
    [x in T]?: string;
  } & { value: string }),
  key: T,
) {
  if (!obj) return undefined;

  return typeof obj === 'string'
    ? obj
    : {
      [`@${key}`]: obj[key],
      '#text': obj.value,
    };
}

export function makeExtensionId(bundleId: string, extensionId: string): string {
  return extensionId.startsWith(bundleId) ? extensionId : `${bundleId}.${extensionId}`;
}

export function objectToXml<K extends string | number, V>(
  obj: Record<K, V> | undefined,
  mapFn: (key: K, value: V) => Record<string, unknown>,
): Record<string, unknown>[] | undefined {
  if (!obj) return undefined;
  return Object.entries(obj).map(([key, value]) => mapFn(key as K, value as V));
}

export function vec2xml(vec?: Vec2) {
  if (!vec) return undefined;
  const [width, height] = vec;
  if (width === undefined && height === undefined) return undefined;

  return {
    Height: height?.toString(),
    Width: width?.toString(),
  };
}
