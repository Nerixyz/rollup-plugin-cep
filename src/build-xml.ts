import { Config, ExecutionEnvironment, ExtensionDispatchInfo } from './types';
import { XMLBuilder } from 'xmlbuilder2/lib/interfaces';
import { create } from 'xmlbuilder2';
import { makeExtensionId, makeVersion, objectToXml, optAttrToXml, vec2xml } from './utilities';

export function buildManifestXml({ extensions, executionEnvironment, ...config }: Config): XMLBuilder {
  const doc = {
    ExtensionManifest: {
      '@Version': makeVersion(config.manifestVersion ?? '9.0'),
      '@ExtensionBundleId': config.bundleId,
      '@ExtensionBundleVersion': makeVersion(config.bundleVersion ?? '1.0.0'),
      '@ExtensionBundleName': config.bundleName,
      Author: config.author,
      Contact: config.contact && { '@mailto': config.contact.mailto, '#text': config.contact.value },
      Legal: optAttrToXml(config.legal, 'href'),
      Abstract: optAttrToXml(config.abstract, 'href'),
      ExtensionList: {
        Extension: extensions.map(ext => ({
          '@Id': makeExtensionId(config.bundleId, ext.id),
          '@Version': ext.version && makeVersion(ext.version),
        })),
      },
      ExecutionEnvironment: executionEnvironmentToXml(executionEnvironment),
      DispatchInfoList: {
        Extension:  extensions.map(extension => ({
          '@Id': makeExtensionId(config.bundleId, extension.id),
          HostList: extension.hostList?.map(host => ({
            Host: {
              '@Name': host,
            },
          })),
          DispatchInfo: extension.dispatchInfos.map(dispatchInfoToXml),
          DependencyList: extension.dependencies && {
            Dependency: objectToXml(extension.dependencies, (id, version) => ({
              '@Id': id,
              '@Version': version,
            })),
          },
        })),
      },
    },
  };

  return create({ encoding: 'UTF-8', standalone: false }, doc);
}

export function buildDebugXml({extensions, bundleId}: Config): XMLBuilder | undefined {
  if(!extensions?.some(ex => ex.debug)) return undefined;

  const doc = {
    ExtensionList: {
      Extension: extensions.filter(ex => ex.debug).map(ex => ({
        '@Id': makeExtensionId(bundleId, ex.id),
        HostList: {
          Host: objectToXml(ex.debug, (name, port) => ({
            '@Name': name,
            '@Port': port,
          })),
        },
      })),
    }
  };

  return create({ encoding: 'UTF-8' }, doc);
}

function executionEnvironmentToXml({ hostList, localeList, requiredRuntime }: ExecutionEnvironment) {
  return {
    HostList: objectToXml(hostList, (name, version) => ({
      Host: {
        '@Name': name,
        '@Version': makeVersion(version),
      },
    })),
    LocaleList: localeList && {
      Locale: localeList.map(loc => ({
        '@Code': loc,
      })),
    },
    RequiredRuntimeList: requiredRuntime && {
      RequiredRuntime: {
        '@Name': 'CSXS',
        '@Version': makeVersion(requiredRuntime),
      },
    },
  };
}

function dispatchInfoToXml({ commandLineParameters, ...info }: ExtensionDispatchInfo) {
  return {
    '@Host': info.host,
    Resources: (info.mainPath || info.scriptPath || commandLineParameters) && {
      MainPath: info.mainPath,
      ScriptPath: optAttrToXml(info.scriptPath, 'engine'),
      CEFCommandLine: commandLineParameters && {
        Parameter: commandLineParameters,
      },
    },
    Lifecycle: {
      AutoVisible: (info.autoVisible ?? true)?.toString(),
      StartOn: info.startOn && {
        Event: info.startOn,
      },
    },
    UI: (info.type ||
      info.menu ||
      info.screenPercentage ||
      info.size ||
      info.maxSize ||
      info.minSize ||
      info.icons) && {
      Type: info.type,
      Menu: optAttrToXml(info.menu, 'placement'),
      Geometry: (info.screenPercentage || info.size || info.minSize || info.maxSize) && {
        ScreenPercentage: vec2xml(info.screenPercentage),
        Size: vec2xml(info.size),
        MaxSize: vec2xml(info.maxSize),
        MinSize: vec2xml(info.minSize),
      },
      Icons: info.icons && {
        Icon: objectToXml(info.icons, (key, value) => ({
          '@Type': key,
          '#text': value,
        })),
      },
    },
  };
}
