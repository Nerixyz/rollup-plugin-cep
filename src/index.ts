import { create } from 'xmlbuilder2';
import { Plugin } from 'rollup';
import { Options } from './types';
import { buildDebugXml, buildManifestXml } from './build-xml';
import { importConfig } from './utilities';

export default function cep(input?: Partial<Options>): Plugin {
  const options: Options = {
    configPath: 'cep.config.js',
    outputPath: 'CSXS/manifest.xml',
    debugOutputPath: '.debug',
    emitDebug: true,
    writerOptions: {
      prettyPrint: true,
    },
    ...input,
  };

  return {
    name: 'cep',
    buildStart() {
      this.addWatchFile(options.configPath);
    },
    generateBundle() {
      const config = importConfig(options.configPath);
      const built = buildManifestXml(config);
      this.emitFile({
        fileName: options.outputPath,
        type: 'asset',
        source: built.toString(options.writerOptions),
      });
      if(!options.emitDebug) return;

      const debug = buildDebugXml(config);
      if(!debug) return;

      this.emitFile({
        fileName: options.debugOutputPath,
        type: 'asset',
        source: debug.toString(options.writerOptions),
      });
    },
  };
}
