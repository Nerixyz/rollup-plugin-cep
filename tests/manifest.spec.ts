import { buildManifestXml } from '../src/build-xml';
import { DefaultConfig, FullConfig } from './test.configs';
// @ts-ignore -- no
const xsd = require('ksys-libxmljs2-xsd');

const schema = xsd.parseFile('tests/manifest.xsd');

describe('buildManifestXml', () => {
  it('should build valid xml', () => {
    const content = buildManifestXml(DefaultConfig);
    const xml = content.toString({prettyPrint: true});

    expect(schema.validate(xml)).toBe(null);
    expect(xml).toMatchSnapshot();
  });

  it('should build a full config', () => {
    const content = buildManifestXml(FullConfig);
    const xml = content.toString({prettyPrint: true});

    expect(schema.validate(xml)).toBe(null);
    expect(xml).toMatchSnapshot();
  });
});
