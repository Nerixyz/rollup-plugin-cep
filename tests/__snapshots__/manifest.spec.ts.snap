// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`buildManifestXml should build a full config 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\" standalone=\\"no\\"?>
<ExtensionManifest Version=\\"7.0\\" ExtensionBundleId=\\"de.nerixyz.bundle\\" ExtensionBundleVersion=\\"2.0\\" ExtensionBundleName=\\"My Bundle\\">
  <Author>Nerixyz</Author>
  <Contact mailto=\\"mail@test.com\\">nerixyz.de</Contact>
  <Legal href=\\"https://nerixyz.de\\">Legal stuff</Legal>
  <Abstract href=\\"https://nerixyz.de\\">Abstract</Abstract>
  <ExtensionList>
    <Extension Id=\\"de.nerixyz.bundle.extension1\\" Version=\\"3.45\\"/>
    <Extension Id=\\"de.nerixyz.bundle.extension2\\" Version=\\"3.45\\"/>
  </ExtensionList>
  <ExecutionEnvironment>
    <HostList>
      <Host Name=\\"AEFT\\" Version=\\"14.0\\"/>
    </HostList>
    <LocaleList>
      <Locale Code=\\"de_DE\\"/>
      <Locale Code=\\"en_GB\\"/>
    </LocaleList>
    <RequiredRuntimeList>
      <RequiredRuntime Name=\\"CSXS\\" Version=\\"4.0\\"/>
    </RequiredRuntimeList>
  </ExecutionEnvironment>
  <DispatchInfoList>
    <Extension Id=\\"de.nerixyz.bundle.extension1\\">
      <HostList>
        <Host Name=\\"4.0\\"/>
      </HostList>
      <DispatchInfo Host=\\"dispatch host\\">
        <Resources>
          <MainPath>./index.html</MainPath>
          <ScriptPath>./index.jsx</ScriptPath>
          <CEFCommandLine>
            <Parameter>--enable-nodejs</Parameter>
            <Parameter>--some-other-parameter</Parameter>
          </CEFCommandLine>
        </Resources>
        <Lifecycle>
          <AutoVisible>false</AutoVisible>
          <StartOn>
            <Event>myEvent</Event>
            <Event>otherEvent</Event>
          </StartOn>
        </Lifecycle>
        <UI>
          <Type>Panel</Type>
          <Menu>My Menu Item</Menu>
          <Geometry>
            <ScreenPercentage>
              <Height>456</Height>
              <Width>123</Width>
            </ScreenPercentage>
            <Size>
              <Height>8</Height>
              <Width>8</Width>
            </Size>
            <MaxSize>
              <Height>9</Height>
              <Width>9</Width>
            </MaxSize>
            <MinSize>
              <Height>7</Height>
              <Width>7</Width>
            </MinSize>
          </Geometry>
          <Icons>
            <Icon Type=\\"DarkNormal\\">./myIcon.webp</Icon>
            <Icon Type=\\"DarkRollOver\\">./myOtherIcon.avif</Icon>
            <Icon Type=\\"Disabled\\">./someOtherIcon</Icon>
            <Icon Type=\\"Normal\\">./something</Icon>
            <Icon Type=\\"RollOver\\">./icon</Icon>
          </Icons>
        </UI>
      </DispatchInfo>
      <DispatchInfo Host=\\"dispatch host\\">
        <Resources>
          <MainPath>./index.html#other-location</MainPath>
          <ScriptPath>./index.jsx</ScriptPath>
          <CEFCommandLine>
            <Parameter>--enable-nodejs</Parameter>
            <Parameter>--some-other-parameter</Parameter>
          </CEFCommandLine>
        </Resources>
        <Lifecycle>
          <AutoVisible>false</AutoVisible>
          <StartOn>
            <Event>myEvent</Event>
            <Event>otherEvent</Event>
          </StartOn>
        </Lifecycle>
        <UI>
          <Type>Panel</Type>
          <Menu>My Second Item</Menu>
          <Geometry>
            <ScreenPercentage>
              <Height>456</Height>
              <Width>123</Width>
            </ScreenPercentage>
            <Size>
              <Height>8</Height>
              <Width>8</Width>
            </Size>
            <MaxSize>
              <Height>9</Height>
              <Width>9</Width>
            </MaxSize>
            <MinSize>
              <Height>7</Height>
              <Width>7</Width>
            </MinSize>
          </Geometry>
          <Icons>
            <Icon Type=\\"DarkNormal\\">./myIcon.webp</Icon>
            <Icon Type=\\"DarkRollOver\\">./myOtherIcon.avif</Icon>
            <Icon Type=\\"Disabled\\">./someOtherIcon</Icon>
            <Icon Type=\\"Normal\\">./something</Icon>
            <Icon Type=\\"RollOver\\">./icon</Icon>
          </Icons>
        </UI>
      </DispatchInfo>
      <DependencyList>
        <Dependency Id=\\"MyDep\\" Version=\\"3\\"/>
      </DependencyList>
    </Extension>
    <Extension Id=\\"de.nerixyz.bundle.extension2\\">
      <HostList>
        <Host Name=\\"4.0\\"/>
      </HostList>
      <DispatchInfo Host=\\"dispatch host\\">
        <Resources>
          <MainPath>./index.html</MainPath>
          <ScriptPath>./index.jsx</ScriptPath>
          <CEFCommandLine>
            <Parameter>--enable-nodejs</Parameter>
            <Parameter>--some-other-parameter</Parameter>
          </CEFCommandLine>
        </Resources>
        <Lifecycle>
          <AutoVisible>false</AutoVisible>
          <StartOn>
            <Event>myEvent</Event>
            <Event>otherEvent</Event>
          </StartOn>
        </Lifecycle>
        <UI>
          <Type>Panel</Type>
          <Menu>My Menu Item</Menu>
          <Geometry>
            <ScreenPercentage>
              <Height>456</Height>
              <Width>123</Width>
            </ScreenPercentage>
            <Size>
              <Height>8</Height>
              <Width>8</Width>
            </Size>
            <MaxSize>
              <Height>9</Height>
              <Width>9</Width>
            </MaxSize>
            <MinSize>
              <Height>7</Height>
              <Width>7</Width>
            </MinSize>
          </Geometry>
          <Icons>
            <Icon Type=\\"DarkNormal\\">./myIcon.webp</Icon>
            <Icon Type=\\"DarkRollOver\\">./myOtherIcon.avif</Icon>
            <Icon Type=\\"Disabled\\">./someOtherIcon</Icon>
            <Icon Type=\\"Normal\\">./something</Icon>
            <Icon Type=\\"RollOver\\">./icon</Icon>
          </Icons>
        </UI>
      </DispatchInfo>
      <DispatchInfo Host=\\"dispatch host\\">
        <Resources>
          <MainPath>./index.html#other-location</MainPath>
          <ScriptPath>./index.jsx</ScriptPath>
          <CEFCommandLine>
            <Parameter>--enable-nodejs</Parameter>
            <Parameter>--some-other-parameter</Parameter>
          </CEFCommandLine>
        </Resources>
        <Lifecycle>
          <AutoVisible>false</AutoVisible>
          <StartOn>
            <Event>myEvent</Event>
            <Event>otherEvent</Event>
          </StartOn>
        </Lifecycle>
        <UI>
          <Type>Panel</Type>
          <Menu>My Second Item</Menu>
          <Geometry>
            <ScreenPercentage>
              <Height>456</Height>
              <Width>123</Width>
            </ScreenPercentage>
            <Size>
              <Height>8</Height>
              <Width>8</Width>
            </Size>
            <MaxSize>
              <Height>9</Height>
              <Width>9</Width>
            </MaxSize>
            <MinSize>
              <Height>7</Height>
              <Width>7</Width>
            </MinSize>
          </Geometry>
          <Icons>
            <Icon Type=\\"DarkNormal\\">./myIcon.webp</Icon>
            <Icon Type=\\"DarkRollOver\\">./myOtherIcon.avif</Icon>
            <Icon Type=\\"Disabled\\">./someOtherIcon</Icon>
            <Icon Type=\\"Normal\\">./something</Icon>
            <Icon Type=\\"RollOver\\">./icon</Icon>
          </Icons>
        </UI>
      </DispatchInfo>
      <DependencyList>
        <Dependency Id=\\"MyDep\\" Version=\\"3\\"/>
      </DependencyList>
    </Extension>
  </DispatchInfoList>
</ExtensionManifest>"
`;

exports[`buildManifestXml should build valid xml 1`] = `
"<?xml version=\\"1.0\\" encoding=\\"UTF-8\\" standalone=\\"no\\"?>
<ExtensionManifest Version=\\"7.0\\" ExtensionBundleId=\\"de.nerixyz.build\\" ExtensionBundleVersion=\\"1.0.0\\">
  <ExtensionList>
    <Extension Id=\\"de.nerixyz.build.test\\" Version=\\"2.0.1\\"/>
  </ExtensionList>
  <ExecutionEnvironment>
    <HostList>
      <Host Name=\\"AEFT\\" Version=\\"14\\"/>
    </HostList>
    <LocaleList>
      <Locale Code=\\"All\\"/>
    </LocaleList>
    <RequiredRuntimeList>
      <RequiredRuntime Name=\\"CSXS\\" Version=\\"4.0\\"/>
    </RequiredRuntimeList>
  </ExecutionEnvironment>
  <DispatchInfoList>
    <Extension Id=\\"de.nerixyz.build.test\\">
      <DispatchInfo>
        <Resources>
          <MainPath>./index.html#root</MainPath>
        </Resources>
        <Lifecycle>
          <AutoVisible>true</AutoVisible>
        </Lifecycle>
        <UI>
          <Type>Panel</Type>
          <Menu>My Menu Item</Menu>
          <Geometry>
            <Size>
              <Height>150</Height>
              <Width>200</Width>
            </Size>
            <MinSize>
              <Height>100</Height>
              <Width>100</Width>
            </MinSize>
          </Geometry>
        </UI>
      </DispatchInfo>
    </Extension>
  </DispatchInfoList>
</ExtensionManifest>"
`;
